core/container.go

```diff
@@ -0,0 +1,388 @@
+		Source: metaSt.ToPB(),
+		Target: metaMount,
+	})
+}
+
+func (s *containerSchema) exitCode(ctx *router.Context, parent *Container, args any) (*int, error) {
+	_, _, mounts, err := parent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mnt := range mounts {
+		if mnt.Target == metaMount {
+			defop, err := llb.NewDefinitionOp(mnt.Source)
+			if err != nil {
+				return nil, err
+			}
+
+			file, err := NewFile(ctx, llb.NewState(defop), "exitCode")
+			if err != nil {
+				return nil, err
+			}
+
+			content, err := file.Contents(ctx, s.gw)
+			if err != nil {
+				return nil, err
+			}
+
+			exitCode, err := strconv.Atoi(string(content))
+			if err != nil {
+				return nil, err
+			}
+
+			return &exitCode, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (s *containerSchema) stdout(ctx *router.Context, parent *Container, args any) (*File, error) {
+	_, _, mounts, err := parent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mnt := range mounts {
+		if mnt.Target == metaMount {
+			defop, err := llb.NewDefinitionOp(mnt.Source)
+			if err != nil {
+				return nil, err
+			}
+
+			return NewFile(ctx, llb.NewState(defop), "stdout")
+		}
+	}
+
+	return nil, nil
+}
+
+func (s *containerSchema) stderr(ctx *router.Context, parent *Container, args any) (*File, error) {
+	_, _, mounts, err := parent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mnt := range mounts {
+		if mnt.Target == metaMount {
+			defop, err := llb.NewDefinitionOp(mnt.Source)
+			if err != nil {
+				return nil, err
+			}
+
+			return NewFile(ctx, llb.NewState(defop), "stderr")
+		}
+	}
+
+	return nil, nil
+}
+
+type containerWithWorkdirArgs struct {
+	Path string
+}
+
+func (s *containerSchema) withWorkdir(ctx *router.Context, parent *Container, args containerWithWorkdirArgs) (*Container, error) {
+	st, cfg, mounts, err := parent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.WorkingDir = args.Path
+
+	return NewContainer(ctx, st, cfg, mounts...)
+}
+
+func (s *containerSchema) workdir(ctx *router.Context, parent *Container, args containerWithVariableArgs) (string, error) {
+	_, cfg, _, err := parent.Decode()
+	if err != nil {
+		return "", err
+	}
+
+	return cfg.WorkingDir, nil
+}
+
+type containerWithVariableArgs struct {
+	Name  string
+	Value string
+}
+
+func (s *containerSchema) withVariable(ctx *router.Context, parent *Container, args containerWithVariableArgs) (*Container, error) {
+	st, cfg, mounts, err := parent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	// NB(vito): buildkit handles replacing properly when we do llb.AddEnv, but
+	// we want to replace it here anyway because someone might publish the image
+	// instead of running it. (there's a test covering this!)
+	newEnv := []string{}
+	prefix := args.Name + "="
+	for _, env := range cfg.Env {
+		if !strings.HasPrefix(env, prefix) {
+			newEnv = append(newEnv, env)
+		}
+	}
+
+	newEnv = append(newEnv, fmt.Sprintf("%s=%s", args.Name, args.Value))
```

Comment:
Not blocking at all, just a thought: is my understanding correct that we are forced into this format of `k=v` because we are re-using the `specs.ImageConfig` struct? I almost wonder if we should just make our own struct that has corresponding fields, but in a format that might be more convenient to us (i.e. `map[string]string` for env).

It's a little copy-paste up front but could make life easier in the long run. Not sure though, whatever you feel is best.

Reply:
That's one reason, but I think it's also worth keeping this a list instead of a map just to be 100% sure we're setting env deterministically. Go will stabilize marshal order, but if we were to loop over a `map[string]string` and do `llb.AddEnv` we'd end up with erroneous cache busting since iteration order is nondeterministic. Splitting on `=` isn't too painful anyhow now that Go as `strings.Cut`.